src/Uno.UI/UI/Xaml/UIElement.Pointers.wasm.cs

```diff
@@ -233,13 +244,25 @@ public static void OnNativePointerEvent()
 	private static PointerRoutedEventArgs ToPointerArgs(
 		UIElement snd,
 		NativePointerEventArgs args,
-		bool? isInContact = null,
 		(bool isHorizontalWheel, double delta) wheel = default)
 	{
+		const int cancel = (int)NativePointerEvent.pointercancel;
+		const int exitOrUp = (int)(NativePointerEvent.pointerout | NativePointerEvent.pointerup);
+
 		var pointerId = (uint)args.pointerId;
+		var pointerType = (PointerDeviceType)args.deviceType;
 		var src = GetElementFromHandle(args.srcHandle) ?? (UIElement)snd;
 		var position = new Point(args.x, args.y);
-		var pointerType = ConvertPointerTypeString(args.typeStr);
+		var isInContact = args.buttons != 0;
+		var isInRange = args.Event is cancel ? false
+			: (args.Event & exitOrUp) != 0 ? pointerType switch
+				{
+					PointerDeviceType.Mouse => true, // Mouse is always in range (unless for 'cancel' eg. if it was unplugged from computer)
+					PointerDeviceType.Touch => false, // If we get a pointer out for touch, it means that pointer left the screen (pt up - a.k.a. Implicit capture)
+					PointerDeviceType.Pen => args.hasRelatedTarget, // If the relatedTarget is null it means pointer left the detection range ... but only for out event!
+					_ => !isInContact // Safety!
+				}
+			: true;
```

Comment:
Nesting ternaries is difficult to read, maybe use nested switches instead?

Reply:
I tried to, but as I check multiple flags at once (`exitOrUp`) I cannot convert it as `switch` statement :(